File: backend/app/core/win_checker.py
# win_checker.py
# All win-condition checks.
# Extracted from the original monolithic engine.py.


# ================= 5 IN A LINE =================

# Canonical direction pairs (each covers both signs automatically)
_LINE_DIRS = [(1, 0), (0, 1)]        # horizontal + vertical
_DIAG_DIRS = [(1, 1), (1, -1)]       # both diagonals
_ALL_DIRS_5 = _LINE_DIRS + _DIAG_DIRS


def check_5_line(board, r, c, player, directions, grid_size, selected_patterns=None):
    """
    Check whether placing at (r, c) completes a 5-in-a-line for player.
    When selected_patterns is provided, only checks directions for patterns that are selected:
      - "LINE"     → horizontal and vertical
      - "DIAGONAL" → both diagonal axes
    When selected_patterns is None (legacy), checks all 4 axis directions.
    Returns (True, line_coords) on win, (False, []) otherwise.
    """
    if selected_patterns is None:
        active = _ALL_DIRS_5
    else:
        active = []
        if "LINE" in selected_patterns:
            active.extend(_LINE_DIRS)
        if "DIAGONAL" in selected_patterns:
            active.extend(_DIAG_DIRS)
        if not active:
            return False, []

    for dr, dc in active:

        count = 1
        line  = [(r, c)]

        for sign in (1, -1):
            rr = r + sign * dr
            cc = c + sign * dc

            while (0 <= rr < grid_size and
                   0 <= cc < grid_size and
                   board[rr][cc] == player):

                count += 1
                line.append((rr, cc))
                rr += sign * dr
                cc += sign * dc

        if count >= 5:
            return True, line

    return False, []


# ================= STRUCTURAL PATTERNS =================

def check_structural_patterns(board, player, patterns, grid_size):
    """
    Check whether player has completed any of the pre-generated shape patterns.
    Returns (True, coords) on match, (False, []) otherwise.
    """
    for pattern in patterns:

        max_r = max(r for r, _ in pattern)
        max_c = max(c for _, c in pattern)

        for br in range(grid_size - max_r):
            for bc in range(grid_size - max_c):

                coords = []
                valid  = True

                for dr, dc in pattern:
                    rr = br + dr
                    cc = bc + dc

                    if board[rr][cc] != player:
                        valid = False
                        break

                    coords.append((rr, cc))

                if valid:
                    return True, coords

    return False, []


def find_path(board, player, target_len, directions, grid_size):
    """
    DFS search for a simple path of target_len connected cells belonging to player.
    Returns the path list if found, None otherwise.
    """
    memo = {}

    def dfs(r, c, path_set):
        if len(path_set) == target_len:
            return list(path_set)
        
        # We don't really need a full memo for "any path of length N" 
        # because the state includes the path_set (visited nodes in current path).
        # But for depth 10-15 on 5x5/7x7, simple backtracking is fine.
        
        for dr, dc in directions:
            nr, nc = r + dr, c + dc
            if (0 <= nr < grid_size and 0 <= nc < grid_size and 
                board[nr][nc] == player and (nr, nc) not in path_set):
                
                path_set.add((nr, nc))
                res = dfs(nr, nc, path_set)
                if res: return res
                path_set.remove((nr, nc))
        return None

    for r in range(grid_size):
        for c in range(grid_size):
            if board[r][c] == player:
                start_set = {(r, c)}
                res = dfs(r, c, start_set)
                if res: return res
    return None

# ================= LARGEST CONNECTED COMPONENT =================

def find_largest_component(board, player, directions, grid_size):
    """
    Flood-fill search for the largest contiguous blob belonging to player.
    Returns the maximum size found and the path list of that component.
    """
    visited = set()
    best_path = []
    max_size = 0

    for r in range(grid_size):
        for c in range(grid_size):
            if board[r][c] == player and (r, c) not in visited:
                queue = [(r, c)]
                visited.add((r, c))
                component_path = [(r, c)]

                while queue:
                    r0, c0 = queue.pop(0)

                    for dr, dc in directions:
                        nr = r0 + dr
                        nc = c0 + dc
                        if (0 <= nr < grid_size and 0 <= nc < grid_size and
                                board[nr][nc] == player and (nr, nc) not in visited):
                            visited.add((nr, nc))
                            queue.append((nr, nc))

                if len(component_path) > max_size:
                    max_size = len(component_path)
                    best_path = component_path

    return max_size, best_path

# ================= FULL-BOARD RESOLUTION =================

def resolve_full_board(board, directions, grid_size):
    """
    Called when the board is completely filled.
    Returns (winner_string, winner_line, p1_score, p2_score).
    Note: p1_score/p2_score are now 10 if path found, 0 otherwise (for simplicity).
    """
    p1_path = find_path(board, "P1", 10, directions, grid_size)
    p2_path = find_path(board, "P2", 10, directions, grid_size)

    p1_score = 10 if p1_path else 0
    p2_score = 10 if p2_path else 0

    if p1_path and not p2_path:
        return "P1", p1_path, p1_score, p2_score
    elif p2_path and not p1_path:
        return "P2", p2_path, p1_score, p2_score
    else:
        # Draw if both have path or neither has path
        return "DRAW", [], p1_score, p2_score

File: frontend/lib/winChecker.ts
// lib/winChecker.ts
// Exact TypeScript port of win_checker.py + patterns.py

export type Coord = [number, number];
export type Board = (string | null)[][];

const GRID = 5;

const DIRS: Coord[] = [
  [-1, 0], [1, 0], [0, -1], [0, 1],
  [-1, -1], [-1, 1], [1, -1], [1, 1],
];

// Named shape patterns only — LINE and DIAGONAL are handled by check5Line
const BASE_PATTERNS_5: Record<string, Coord[]> = {
  V:      [[0, 0], [1, 1], [2, 2], [1, 3], [0, 4]],
  L:      [[0, 0], [1, 0], [2, 0], [2, 1], [2, 2]],
  "ZZ-5": [[0, 0], [1, 1], [2, 0], [3, 1], [4, 0]],
  T:      [[0, 0], [0, 1], [0, 2], [1, 1], [2, 1]],
};

const LINE_DIRS: Coord[] = [[1, 0], [0, 1]];       // horizontal + vertical
const DIAG_DIRS: Coord[] = [[1, 1], [1, -1]];      // both diagonals

// ── Variant Generation ──
function generateVariants(pattern: Coord[]): Coord[][] {
  const variants = new Set<string>();
  const result: Coord[][] = [];

  for (const reflect of [1, -1]) {
    for (let rotation = 0; rotation < 4; rotation++) {
      const transformed: Coord[] = [];
      for (const [r, c] of pattern) {
        let r2 = r, c2 = c * reflect;
        for (let i = 0; i < rotation; i++) {
          [r2, c2] = [c2, -r2];
        }
        transformed.push([r2, c2]);
      }
      const minR = Math.min(...transformed.map(([r]) => r));
      const minC = Math.min(...transformed.map(([, c]) => c));
      const normalized: Coord[] = transformed
        .map(([r, c]) => [r - minR, c - minC] as Coord)
        .sort((a, b) => a[0] - b[0] || a[1] - b[1]);
      const key = normalized.map(([r, c]) => `${r},${c}`).join("|");
      if (!variants.has(key)) {
        variants.add(key);
        result.push(normalized);
      }
    }
  }
  return result;
}

export const ALL_VARIANTS_5: Record<string, Coord[][]> = {};
for (const name in BASE_PATTERNS_5) {
  ALL_VARIANTS_5[name] = generateVariants(BASE_PATTERNS_5[name]);
}

// ─── 5-in-a-line ───
export function check5Line(board: Board, r: number, c: number, player: string, selectedPatterns?: string[]): Coord[] | null {
  let activeDirs: Coord[];
  if (!selectedPatterns || selectedPatterns.length === 0) {
    activeDirs = [...LINE_DIRS, ...DIAG_DIRS];
  } else {
    activeDirs = [];
    if (selectedPatterns.includes("LINE")) activeDirs.push(...LINE_DIRS);
    if (selectedPatterns.includes("DIAGONAL")) activeDirs.push(...DIAG_DIRS);
    if (activeDirs.length === 0) return null;
  }
  for (const [dr, dc] of activeDirs) {
    const line: Coord[] = [[r, c]];
    for (const sign of [1, -1] as const) {
      let rr = r + sign * dr;
      let cc = c + sign * dc;
      while (rr >= 0 && rr < GRID && cc >= 0 && cc < GRID && board[rr][cc] === player) {
        line.push([rr, cc]);
        rr += sign * dr;
        cc += sign * dc;
      }
    }
    if (line.length >= 5) return line;
  }
  return null;
}

// ─── Structural patterns ───
export function checkStructuralPatterns(board: Board, player: string, activeIds: string[]): Coord[] | null {
  const patterns: Coord[][] = [];
  for (const id of activeIds) {
    if (ALL_VARIANTS_5[id]) {
      patterns.push(...ALL_VARIANTS_5[id]);
    }
  }

  for (const pattern of patterns) {
    const maxR = Math.max(...pattern.map(([r]) => r));
    const maxC = Math.max(...pattern.map(([, c]) => c));
    for (let br = 0; br < GRID - maxR; br++) {
      for (let bc = 0; bc < GRID - maxC; bc++) {
        const coords: Coord[] = [];
        let valid = true;
        for (const [dr, dc] of pattern) {
          const rr = br + dr, cc = bc + dc;
          if (board[rr][cc] !== player) { valid = false; break; }
          coords.push([rr, cc]);
        }
        if (valid) return coords;
      }
    }
  }
  return null;
}

// ─── Flood Fill: find largest contiguous blob ───
function findPath(board: Board, player: string, targetLen: number): Coord[] | null {
  function dfs(r: number, c: number, path: Coord[], pathSet: Set<string>): Coord[] | null {
    if (path.length === targetLen) return path;

    for (const [dr, dc] of DIRS) {
      const nr = r + dr, nc = c + dc;
      const key = `${nr},${nc}`;
      if (
        nr >= 0 && nr < GRID && nc >= 0 && nc < GRID &&
        board[nr][nc] === player && !pathSet.has(key)
      ) {
        pathSet.add(key);
        path.push([nr, nc]);
        const res = dfs(nr, nc, path, pathSet);
        if (res) return res;
        path.pop();
        pathSet.delete(key);
      }
    }
    return null;
  }

  for (let r = 0; r < GRID; r++) {
    for (let c = 0; c < GRID; c++) {
      if (board[r][c] === player) {
        const res = dfs(r, c, [[r, c]], new Set([`${r},${c}`]));
        if (res) return res;
      }
    }
  }
  return null;
}

// ─── Full board resolution ───
function resolveFullBoard(board: Board): { winner: string; line: Coord[]; connectionScores?: { p1: number; p2: number } } {
  const p1Path = findPath(board, "P1", 10);
  const p2Path = findPath(board, "P2", 10);
  const scores = { p1: p1Path ? 10 : 0, p2: p2Path ? 10 : 0 };

  if (p1Path && !p2Path) return { winner: "P1", line: p1Path, connectionScores: scores };
  if (p2Path && !p1Path) return { winner: "P2", line: p2Path, connectionScores: scores };
  return { winner: "DRAW", line: [], connectionScores: scores };
}

// ─── Main entry ───
export function checkWin(
  board: Board,
  r: number,
  c: number,
  player: string,
  movesPlayed: number,
  selectedPatternIds: string[] = []
): { winner: string; line: Coord[]; connectionScores?: { p1: number; p2: number } } | null {
  const line5 = check5Line(board, r, c, player, selectedPatternIds.length > 0 ? selectedPatternIds : undefined);
  if (line5) return { winner: player, line: line5 };

  // If no patterns selected (e.g. legacy or not yet initialized), default to standard 5x5 set
  const activeIds = selectedPatternIds.length > 0 ? selectedPatternIds : ["V", "L", "ZZ-5", "T"];
  const lineS = checkStructuralPatterns(board, player, activeIds);
  if (lineS) return { winner: player, line: lineS };

  if (movesPlayed === GRID * GRID) return resolveFullBoard(board);

  return null;
}